### ISSUE TITLE
Deleted topics causing cascading metadata refreshes

### ISSUE THREAD

**Issue by kburdett** [tulios/kafkajs#651]
I have encountered an issue with extensive metadata refreshes when a topic is deleted from the Kafka cluster. The refresh counts are far in excessive of the retry policy.

KakfkaJS 1.12.0
Node 11.15.0

### Symptoms
A producer performs many metadata refreshes, spanning close to a minute, for a single send request. Normally, the retry policy limits this to 5 attempts and the error is returned in a few seconds. However, in the case of a deleted topic, the refresh is attempted many times and introduces a significant delay in response.

### Steps to reproduce
1. Create a producer
2. Send a message to topic X
3. Delete X from the Kafka cluster
4. Send another message to topic X

### What I have uncovered so far
This seems very similar to #252, but the fix there does not correct this particular problem. I think there are really two parts to this problem:

First, it seems that when metadata is fetched for a topic, this topic name is added to the `targetTopics` Set on the cluster object. Future metadata refreshes for this producer include all members of the `targetTopics` Set. So, once _any_ member of the `targetTopics` set is deleted, _all_ metadata refreshes on this producer fail with an "This server does not host this topic-partition" error. 

Second, there does seem to be nested instances of the retrier at play here. That would explain why the 5 retry policy is being violated. In my case, there were 42 errors with message "Response Metadata" and 36 with "Response Produce" taking just over a minute while using the default retry policy. Possibly 5 retries on the send with 5 retries on metadata refresh per send attempt? The numbers don't quite add up, but that could be a timing thing? I'm uncertain.

### Possible solutions
I am happy to make an attempt a PR but I'd like some feedback first, as I don't know the codebase very well.

I thought to simply remove the topic from the `targetTopics` Set when this occurs. This is the solution implemented in #252. I haven't dug into the returned message, but I'm assuming that Kafka doesn't identify which topic is at fault, so this might not be possible when `targetTopics` contains more than one topic.

On a deeper level, I'm not really sure what `targetTopics` is here for. It seems like the producer is "over-refreshing" and fetching metadata for _all topics_ that it knows about, rather than just the topic _currently being acted upon_. If we could avoid this behavior, then a solution would perhaps be easier here. We could possibly clear the `targetTopics` Set when refresh errors are encountered.  I'm not sure what the implications could be here. Possibly `targetTopics` could be eliminated in cases where only a single topic refresh is necessary?

As for the over-retry issue, I've not made much progress here. Any guidance would be appreciated.

**Comment by Nevon** [comment]
I created a branch with a repro case for this. One difficulty is that this only fails if topic auto creation is turned off, which is a broker level config that we can't really change from the client side, so it requires configuring the containers differently. Maybe it's fine, but there might be some tests that need to explicitly create the topic first.

I'll need to get back to you on the questions you had, but at least the issue is verified and we have something to work from.

**Comment by Nevon** [comment]
> On a deeper level, I'm not really sure what targetTopics is here for. It seems like the producer is "over-refreshing" and fetching metadata for all topics that it knows about, rather than just the topic currently being acted upon.

I think this is an interesting idea that would be cool to see if it's possible to implement. That part of the cluster was written 2 years ago, so I don't really recall the circumstances, but my current thinking is that `targetTopics` is just a convenience to make it simple to refresh metadata, but you are right that we could do more targeted refreshes of just the topics that we care about, instead of over-refreshing.

I believe more in that approach than trying to remove the topic from the targetgroup when you get this error, because most likely there are several operations that could fail if the topic doesn't exist, possibly with different errors. Maybe it's worth exploring, but I would consider it plan B.

**Comment by kburdett** [comment]
Thanks for the feedback @Nevon . I think this is possible. I'm working on a PR to do this.

From what I can tell, there are a few functional areas referencing `targetTopics`: 
 * `getTopicMetadata`
 * `findLeaderForPartitions`
 * `addTargetTopic`
 * `deleteTargetTopic`
Adjusting `findLeaderForPartitions` seems to be fairly straightforward, but it will likely be breaking changes for the others.

I was hoping to have something for you to look at yesterday, but I'm struggling a bit in the testing infrastructure. I have (almost) all the tests passing again, and I'm trying to add some new ones to cover new cases. I hope to have it ready today, so I can get some feedback on the approach.

**Comment by Nevon** [comment]
All the methods are internal to KafkaJS, so breaking changes are fine - it's just gonna be a pain to implement.

We're all over in Europe, so even if you get it ready today, it's almost tomorrow here, so you'll have to wait a bit on feedback. Feel free to open up a pr even if there are still places that aren't finished.

A tip for when developing, run scripts/dockerComposeUp.sh and then scripts/createScramCredentials.sh to run the Kafka cluster, then you can run yarn test:local:watch to run the tests without starting and stopping the containers.

**Comment by kburdett** [comment]
On your advise, I opened up a PR and added some comments there for consideration. I'm still digging through the tests, but I think the meat is there.

I haven't tried the watch script yet. I'll give it a shot. I keep getting intermittent timeouts in the tests, and a couple failures that I'm all but convinced are unrelated.

**Comment by kburdett** [comment]
@Nevon I have spent some time digging into the multiple retry issue and found the source. It is, in fact, a nested retrier.

You can see here where the error condition is tripped by the deleted topic:
https://github.com/tulios/kafkajs/blob/master/src/producer/messageProducer.js#L90-L99

This call to `refreshMetadata` is occurring within a retrier.

Inside `brokerPool.refreshMetadata`, you can see that a new retrier instance is being utilized:
https://github.com/tulios/kafkajs/blob/master/src/cluster/brokerPool.js#L125

I did a quick audit of `refreshMetadata` calls and found it used in mixed scenarios, sometimes inside a retrier, and sometimes outside. In any case where `refreshMetadata` fails for a reason that is flagged as retriable, these nested retriers may exceed the retry policy time restrictions.

I have thought of a couple potential solutions, and wanted to get your thoughts.

1. An optional flag could be added to `refreshMetadata` indicating whether or not it should be retried. This would enable the calling code to disable the retry in cases where the calling code itself is already in a retrier.
2. The retrier could be removed from `refreshMetadata`, in the 6 or so cases where it is called outside of one today, the caller could wrap it in a retrier itself. Effects to the public contract could be mitigated by adding a wrapping retrier in the admin module.

**Comment by Nevon** [comment]
The nested retrier inside `brokerPool.refreshMetadata` is only retrying on `LEADER_NOT_AVAILABLE` errors, though. For anything else, it immediately bails out. Granted, if that happens consistently in a nested retrier, you will get very long retry times. That said, if that's happening consistently, there's not much we can do, since we need the leader to be available before being able to do much of anything.

**Comment by kburdett** [comment]
Of course, you are right. I am definitely seeing a retry issue, but it seems I've misidentified the source. I did some more digging... I added a `console.error` to the catch block in `refreshMetadata` just to capture the stack traces. I then sent a single produce request to a non-existent topic. It produced 78 `refreshMetadata` attempts and took well over a minute before failing.

I cataloged all of the exceptions based on their stack trace, and found a pattern. I tracked down three separate locations where the metadata is being refreshed:

*First*
```
producer/messageProducer.js:74
    producer/sendMessages.js:122
        producer/sendMessages.js:26
```
*Second*
```
producer/messageProducer.js:74
    producer/sendMessages.js:128
```
*Third*
```
producer/messageProducer.js:98
```
[Here is the full log of metadata refresh errors, including stack traces, in case you are interested in taking a peek](https://github.com/tulios/kafkajs/files/4411118/refreshMetadata.errors.log)

I believe in this case, the nested retriers are these:
outer: https://github.com/tulios/kafkajs/blob/master/src/producer/messageProducer.js#L71
inner: https://github.com/tulios/kafkajs/blob/master/src/producer/sendMessages.js#L137

In my case, the error being returned by the server is UNKNOWN_TOPIC_OR_PARTITION. This code is categorized as stale metadata and causing the retries inside `sendMessages`. I think, given the code, this is a reasonable response. I fear that removing this code from the stale list and aborting the retries might cause too many issues elsewhere.

The function `sendMessages` appears to only have a single usage, and that is the reference here from `messageProducer`. Given this, it seems safe to remove the retrier here and allow the outer retrier of `messageProducer` handle the retries. I tested this out, it certainly corrects the over-refresh case.

I think this change to eliminate the nested retrier is sound. If you agree, I can cleanup the `sendMessages` tests and incorporate the changes into #667, or create another PR specifically for it. What do you think?

**Comment by Nevon** [comment]
I think creating a new PR for it is better. Makes it easier to review the change.

Just make sure that you handle the same cases that the inner retrier did, now in the outer retrier. We have to make sure to refresh metadata in those cases that the inner retrier is currently handling.

**Comment by kburdett** [comment]
@Nevon I have encountered an issue I'm not certain how to deal with. The `sendMessages` function maintains a map of broker responses. The retrier utilized there only retries the brokers that do not yet have a successful response. Relying on the outer retrier will break this functionality and force a full refresh for each failure. I don't know what this implies, but working around this test would require breaking down a few walls of separation between these two modules. Do you feel like this is worth pursuing further?

**Comment by Nevon** [comment]
Ah, right, that's extremely important to avoid sending duplicates of requests that succeeded, so it's something that has to be worked around somehow.

Just a random thought I have - maybe the problem we're having here is the assumption that `UNKNOWN_TOPIC_OR_PARTITION` should always be classified as stale metadata. Maybe the solution is to bail out if we already refreshed metadata and we still get the same error.

**Comment by wired8** [comment]
@Nevon @kburdett did you guys ever figure out a solution for this issue?  The only hack I could think of for now is to force a re-connect on error to synchronize the topic list.
